tabulon_vello/src/lib.rs

```diff
@@ -162,4 +164,69 @@ impl Environment {
             }
         }
     }
+
+    /// Measure text items in a [`RenderLayer`].
+    #[tracing::instrument(skip_all)]
+    pub fn measure_text_items(
+        &mut self,
+        graphics: &GraphicsBag,
+        render_layer: &RenderLayer,
+    ) -> BTreeMap<ItemHandle, (DirectIsometry, Size)> {
+        let Self { font_cx, layout_cx } = self;
+        let mut out = BTreeMap::new();
+
+        for idx in &render_layer.indices {
+            let Some(GraphicsItem::FatText(FatText {
+                text,
+                style,
+                max_inline_size,
+                alignment,
+                insertion,
+                attachment_point,
+                ..
+            })) = graphics.get(*idx)
+            else {
+                continue;
+            };
+
+            let mut builder = layout_cx.ranged_builder(font_cx, text, 1.0);
+            for prop in style.inner().values() {
+                builder.push_default(prop.to_owned());
+            }
+            let mut layout = builder.build(text);
+            layout.break_all_lines(*max_inline_size);
+            layout.align(*max_inline_size, *alignment, Default::default());
+
+            let layout_size = Size {
+                width: max_inline_size.unwrap_or(layout.width()) as f64,
+                height: layout.height() as f64,
+            };
+
+            let rotated_offset = rotate_offset(*attachment_point, layout_size, insertion.angle);
+
+            out.insert(
+                *idx,
+                (
+                    DirectIsometry {
+                        displacement: insertion.displacement - rotated_offset,
+                        ..*insertion
+                    },
+                    layout_size,
+                ),
+            );
+        }
+
+        out
+    }
+}
+
+/// Calculate a top left equivalent insertion point for a layout size and attachment point.
+fn rotate_offset(attachment_point: AttachmentPoint, layout_size: Size, angle: f64) -> Vec2 {
+    let attachment = attachment_point.select(layout_size);
+    let cos_theta = angle.cos();
+    let sin_theta = angle.sin();
```

Comment:
Can this be `let (sin_theta, cos_theta) = angle.sincos();` ?

Reply:
yep, forgot because I did it piecemeal :+1: